onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.cc

```diff
@@ -338,35 +345,43 @@ Status DQMatMulToMatMulNBitsAction::ProcessNewNode(Graph& graph,
   // to what we need. But it does not handle external data.
   Initializer weight_src(*weight_tensor_proto, graph.ModelPath());
   Initializer scale_src(*scale_tensor_proto, graph.ModelPath());
-  std::optional<Initializer> zp_src;
-  Initializer weight_dst(ONNX_NAMESPACE::TensorProto_DataType_UINT8,
-                         graph.GenerateNodeArgName(weight_arg->Name() + "_T"),
-                         std::vector<int64_t>{N, quant_num, blob_bytes});
-  Initializer scale_dst(static_cast<ONNX_NAMESPACE::TensorProto_DataType>(scale_src.data_type()),
-                        graph.GenerateNodeArgName(scale_arg->Name() + "_T"),
-                        std::vector<int64_t>{N * quant_num});
-  std::optional<Initializer> zp_dst;
+  auto uint8_type = DataTypeImpl::TensorTypeFromONNXEnum(ONNX_NAMESPACE::TensorProto_DataType_UINT8)->GetElementType();
+  auto scale_type = DataTypeImpl::TensorTypeFromONNXEnum(scale_src.data_type())->GetElementType();
+  std::optional<Initializer> zp_src_ptr;
+  auto weight_dst_name = graph.GenerateNodeArgName(weight_arg->Name() + "_T");
+  auto weight_dst_ptr = std::make_unique<Tensor>(uint8_type,
+                                                 TensorShape{N, quant_num, blob_bytes},
+                                                 std::make_shared<CPUAllocator>());
```

Comment:
Instead of creating multiple allocator here, better to pass it in.  Please add a "const AllocatorPtr& alloc" parameter to this function.

Reply:
`Tensor::Init` calls `buffer_deleter_ = std::move(deleter);`. If we pass ref, after first ctor, the alloc created here will be undefine. Each ctor requires its own `AllocatorPtr` object.

Reply:
Some modifications in the Tensor ctor code is required to use shared_ptr + ref. I'd better not touch that part in this PR.

Reply:
We may use the same CPUAllocator, but with different pointers to it.

Reply:
updated